build_projects/dotnet-cli-build/TarGzFileExtractToDirectory.cs

```diff
@@ -0,0 +1,112 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using System;
+using System.IO;
+using System.IO.Compression;
+
+namespace Microsoft.DotNet.Build.Tasks
+{
+    public sealed class TarGzFileExtractToDirectory : ToolTask
+    {
+        /// <summary>
+        /// The path to the archive to extract.
+        /// </summary>
+        [Required]
+        public string SourceArchive { get; set; }
+
+        /// <summary>
+        /// The path of the directory to which the archive should be extracted.
+        /// </summary>
+        [Required]
+        public string DestinationDirectory { get; set; }
+
+        /// <summary>
+        /// Indicates if the destination directory should be cleaned if it already exists.
+        /// </summary>
+        public bool OverwriteDestination { get; set; }
+
+        protected override bool ValidateParameters()
+        {
+            base.ValidateParameters();
+
+            var retVal = true;
+
+            if (Directory.Exists(DestinationDirectory))
+            {
+                if (OverwriteDestination == true)
+                {
+                    Log.LogMessage(MessageImportance.Low, "'{0}' already exists, trying to delete before unzipping...", DestinationDirectory);
+                    Directory.Delete(DestinationDirectory, recursive: true);
+                }
+                else
+                {
+                    Log.LogWarning("'{0}' already exists. Did you forget to set '{1}' to true?", DestinationDirectory, nameof(OverwriteDestination));
+
+                    retVal = false;
+                }
+            }
+
+            if (!File.Exists(SourceArchive))
+            {
+                Log.LogError($"SourceArchive '{SourceArchive} does not exist.");
+
+                retVal = false;
+            }
+
+            if (retVal)
+            {
+                Log.LogMessage($"Creating Directory {DestinationDirectory}");
+                Directory.CreateDirectory(DestinationDirectory);
+            }
+            
+            return retVal;
+        }
+
+        public override bool Execute()
+        {
+            bool retVal = base.Execute();
+
+            if (!retVal)
```

Comment:
If overwrite is set then we explicitly delete the directory before unpacking. Overwrite in this scenario cannot be additive because we don't know what was the set of files previously generated by unzipping this file.

This is a direct port of the logic in [build_tools](https://github.com/dotnet/buildtools/blob/master/src/Microsoft.DotNet.Build.Tasks/ZipFileExtractToDirectory.cs#L40)